apps/s_client.c

```diff
@@ -3553,4 +3556,70 @@ static char *base64encode (const void *buf, size_t len)
     return out;
 }
 
+/*
+ * Host dNS Name verifier: used for checking that the hostname is in dNS format 
+ * before setting it as SNI
+ */
+static int is_dNS_name(const char *host)
+{
+    const int MAX_LABEL_LENGTH = 63;
+
+    int i;
+    int isdnsname = 0;
+    int length = strlen(host);
+    int label_length = 0;
+    int all_numeric = 1;
+
+    /*
+     * XXX: Deviation from strict DNS name syntax, also check names with '_'
+     * Check DNS name syntax, any '-' or '.' must be internal,
+     * and on either side of each '.' we can't have a '-' or '.'.
+     *
+     * If the name has just one label, we don't consider it a DNS name.
+     */
+    for (i = 0; i < length && label_length < MAX_LABEL_LENGTH; ++i) {
+        char c = host[i];
+
+        if ((c >= 'a' && c <= 'z')
+            || (c >= 'A' && c <= 'Z')
```

Comment:
The above lines look like they're going to fail on an EBCDIC system. Probably you're going to need to use `os_toascii` if `CHARSET_EBCDIC` is defined. Have a look elsewhere in the code for examples.

Reply:
Does that mean cn2dnsid is potentially an issue as well? (asking out of curiosity)
I know that one does some conversion before hand, though - from ASN.1 to UTF8 I think? Might be talking rubbish here

Reply:
Yes - actually looking at cn2dnsid I suspect it could be an issue. @vdukhovni?

Reply:
Actually thinking about this again it might be the case that I'm wrong here since we're comparing two values both of which would be in the same character set - so as long as the letters are in a contiguous range for EBCDIC (which seems to be more-or-less true) then it's probably ok after all. OTOH cn2dnsid could still be a problem if one side of the comparison has been converted to UTF8 first (7 bit ASCII is a subset of UTF8, but EBCDIC isn't).

Reply:
In `cn2dnsid` we're trying to determine whether a `CommonName` element of a subject name from an X.509 certificate is plausibly a DNS name. The `CommonName` is not an EBCDIC string. When converted to UTF-8 it is still not an EBCDIC string, but rather a UTF-8 encoded byte array.  So the only question is whether the C compiler is working with EBCDIC character constants and ordering.

The `cn2dnsid` code has been in place since 1.0.2 and nobody has complained, perhaps IBM have their own patches?  But if we wanted to address this the data type of the UTF-8 buffer could be changed to 'unsigned int8 *' and the comparisons could be with hexadecimal constants for:
```
'a' == 0x61
'A' == 0x41
'0' == 0x30
'9' == 0x39
'z' == 0x7a
'Z' == 0x5a
```
And then also `'-'` and `'.'` would have to use the corresponding hex constants.

Reply:
As for the hostname, where is the user's input string being converted from EBCDIC for use with SNI?  Does anyone still care about EBCDIC in this context?

Reply:
> So the only question is whether the C compiler is working with EBCDIC character constants and ordering.

Yes, this was the aspect I was concerned about.

> The `cn2dnsid` code has been in place since 1.0.2 and nobody has complained, perhaps IBM have their own patches? But if we wanted to address this the data type of the UTF-8 buffer could be changed to 'unsigned int8 *' and the comparisons could be with hexadecimal constants for:

That's what `os_toascii` does for you.

> As for the hostname, where is the user's input string being converted from EBCDIC for use with SNI? Does anyone still care about EBCDIC in this context?

It isn't. I suspect this is yet another bug. My understanding is that OpenSSL is still used on EBCDIC systems (and we certainly have code for it in the codebase). But we don't have access to one to perform tests :-(

Reply:
Coming back to the method in this PR - should I convert the hostname and characters in the comparison to ASCII before/when using them? It looks like just comparing should work('ish), but it could allow characters that it shouldn't - only '~' as far as I can tell.

Reply:
I think it's probably ok as is (sorry for the distraction).

Reply:
I'd be ok making a stab at updating cn2dnsid if you consider it necessary - it's probably better suited for a new issue, but I can pick that up after this one goes through.

Reply:
Don't know what you have in mind for `cn2dnsid`, but the real EBCDIR issues are not so much there, as in the functions that take strings from the application for comparison with the certificate.  What character set are those strings in?  We don't know.  So probably best to assume ASCII, with the application responsible for any conversion to make it so.  The main change that might be sound is to use `os_toascii['A'], os_toascii['Z']` and similar instead of the character literals, because the compiler's charset might not be ASCII.

Reply:
Yeah, that's what I was thinking of changing. Whether it's needed is your decision, obviously :) was just saying that I'd be happy to follow through if you decide it's good to have

Reply:
Personally I think the change should be made.

Reply:
Perhaps so, but I'm concerned that we may be appearing to promise more EBCDIC support than is really there. The change of course is fine, and probably should be made, and will do no harm, but is likely insufficient to actually result in working name checks on EBCDIC systems.  A more comprehensive patch needs to come from someone supporting an EBCDIC platform who can perform meaningful tests.

Reply:
Should I open a new issue?

Reply:
You can do. Or just create the PR.

Reply:
There's nothing to do on this request anymore, but I'll have a look and maybe I can identify how many things would need changing for EBCDIC. But as @vdukhovni said, without being able to properly test it don't know how dangerous the changes would be

Reply:
Making sure that character constants are used portably when comparing with wire octet strings should be generally safe.  Converting application input to ASCII risks breakage, because applications may have (should/would have given the prior implementation) already done that.  So there's only so much we can do, but a few judicious `os_toascii()` uses would be OK, though I'd be tempted to just use hex constants, we know what the ASCII codes for `0, 9, a, A, z, Z` are!  So for strings whose character set we already know (UTF-8 form of CommonName, is a case in point), we can just view them is int8 arrays and use hex values for the upper/lower bounds of character ranges.

But I'd leave this alone in the stable releases.  Nobody is complaining.  Just something to consider for "master" (a.k.a. 3.0.0).